rust/parquet/src/column/reader.rs

```diff
@@ -90,21 +89,11 @@ pub fn get_column_reader(
 /// Gets a typed column reader for the specific type `T`, by "up-casting" `col_reader` of
 /// non-generic type to a generic column reader type `ColumnReaderImpl`.
 ///
-/// NOTE: the caller MUST guarantee that the actual enum value for `col_reader` matches
-/// the type `T`. Otherwise, disastrous consequence could happen.
+/// Panics if actual enum value for `col_reader` does not match the type `T`.
 pub fn get_typed_column_reader<T: DataType>(
     col_reader: ColumnReader,
 ) -> ColumnReaderImpl<T> {
-    match col_reader {
-        ColumnReader::BoolColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::Int32ColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::Int64ColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::Int96ColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::FloatColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::DoubleColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::ByteArrayColumnReader(r) => unsafe { mem::transmute(r) },
-        ColumnReader::FixedLenByteArrayColumnReader(r) => unsafe { mem::transmute(r) },
-    }
+    T::get_column_reader(col_reader).expect("Column type mismatch")
```

Comment:
Well, your fix does not change the outcome - we can still fail by passing the wrong type, except we would directly panic now. The original function implementation is actually "safe" despite using `unsafe` - there is a strict list of physical types that is allowed to be used.

I would probably return an Option<ColumnReaderImpl<T>> instead. Same for writing.

Reply:
> Your fix does not change the outcome - we can still fail by passing the wrong type, except we would directly panic now.

Sure, but that is much better than undefined behaviour.

> The original function implementation is actually "safe" despite using unsafe - there is a strict list of physical types that are allowed to be used.

That is not enough to make it safe. To transmute a `ColumnReader<A>` into a `ColumnReader<B>`  the representations of `A` and `B` must be identical. `A` and `B` are both empty structs in this case https://github.com/apache/arrow/blob/a9ef8280524ada2857d1554bc0688e15af4e5404/rust/parquet/src/data_type.rs#L405 which may be enough to make Rust lay them out in the same way but I wouldn't rely on it without any confirmation from the reference.

In any case it can be implemented with safe code which also has the added benefit of actually panicking if you do the wrong thing instead of continuing silently.

> Actually, I would probably return an Option<ColumnReaderImpl> instead. Same for writing.

That would force callers to unwrap instead which seems redundant. Propagating the error seems unnecessarily complicated as well as it should just be a programming error (should not be visible to users of the crate).

Reply:
We never transmute ColumnReader<A> into ColumnReader<B>, so the rest is questionable as well.

Reply:
Yeah, but mistakes happen and now it will panic instead of causing undefined behaviour.

Reply:
* there is no "we never do X" in a public interface.
* the code should immediately panic if someone uses it in the wrong way. This should not be an `Option` because it is a programing error.
* failing as undefined bahavior is worse than failing by panic. (why are even having this discussion in rust...)
* changing the return type is a backward incompatible change... WTF?